### DISCUSSION TITLE
syncfinder

### DISCUSSION THREAD

**Discussion by bai-0829** [swesterfeld/audiowmark#42] [Q&A]
Hello, first of all, thank you for your code, but I have a few questions to ask you. 1. In the code, I found that two arrays, up and down, were created, and I am not very clear about their purpose. 2. How does the system find the beginning of the 510 frame synchronization block after cropping the audio? In other words, how does the system find the beginning of the watermark after cropping the audio.

**Comment by swesterfeld** [comment]
I've converted this to discussions. I'm planning to provide better developer documentation, but I don't have the time to do it right now.

> 1. In the code, I found that two arrays, up and down, were created, and I am not very clear about their purpose.

The basic algorithm audiowmark uses for watermarking is the patchwork algorithm to embed information in the amplitudes of the FFT. The up and down arrays are used to determine which frequency bands of a frame are used to embed the bit. I've mentioned a link to a paper that at least briefly describes the algorithm here: https://github.com/swesterfeld/audiowmark/issues/27


> 2. How does the system find the beginning of the 510 frame synchronization block after cropping the audio? In other words, how does the system find the beginning of the watermark after cropping the audio.

The basic idea here is that each block contains data bits which carry the actual watermarking data and sync bits. The value of the data bits of a watermarked block are not known and depend on the message bits that were embedded. However the values of the sync bits are known. So what we can do to find the position of the watermark is this:

- move start position over the whole audio file
- for each start position decode the sync bits
- if the sync bits contain the "right" value, assume that this is the position of the watermark, so decode the data bits
